File: functions/consistency/app/calculator/calculator.py
from typing import Sequence, TypeAlias
import logging
import random
from collections import Counter
from typing import Callable, Sequence, TypeAlias, TypedDict, Union, NotRequired

from app.calculator.exceptions import InvalidCardCountsError
from app.calculator.result import ConsistencyResult, HandTestResult

logger = logging.getLogger()
logger.setLevel("INFO")

CardID: TypeAlias = int
AttrKey: TypeAlias = tuple[str, str]
ExactPattern: TypeAlias = dict[CardID, int]
WildPattern: TypeAlias = dict[AttrKey, int]
CompiledPattern: TypeAlias = tuple[ExactPattern, WildPattern]
CardAttrIndex: TypeAlias = dict[CardID, dict[AttrKey, int]]


class Card(TypedDict):
    superType: str
    name: str
    race: NotRequired[str]
    attribute: NotRequired[str]


CardDatabase: TypeAlias = dict[CardID, Card]
DiscardConstraint: TypeAlias = tuple[str, str]


class GambleCard(TypedDict):
    draw: int
    discard: NotRequired[list[DiscardConstraint]]


GamblingCards: TypeAlias = dict[CardID, GambleCard]


def hand_is_good(
    hand: Sequence[int],
    ideal_hands: Sequence[Union[Sequence[int | str], Counter]]
) -> bool:
    """
    Simple checking from original code, no longer called.
    Return True if the hand matches any of the ideal hands.
    Accepts ideal_hands as list of lists (for tests) or list of Counters (optimized).
    """
    hand_counter = Counter(hand)

    # Convert ideal_hands to counters if not already
    ideal_counters = [
        pattern if isinstance(pattern, Counter) else Counter(pattern)
        for pattern in ideal_hands
    ]

    return any(
        all(hand_counter[card] >= count for card, count in pattern.items())
        for pattern in ideal_counters
    )


def hand_is_wild(
    hand: Sequence[CardID],
    compiled_patterns: list[CompiledPattern],
    card_attr_index: CardAttrIndex,
) -> int:
    """
    Return the number of compiled patterns matched by the hand.
    """
    hand_counter = {}
    for c in hand:
        hand_counter[c] = hand_counter.get(c, 0) + 1

    attr_counter = {}
    for card in hand:
        card_attrs = card_attr_index.get(card)
        if card_attrs:
            for attr, val in card_attrs.items():
                attr_counter[attr] = attr_counter.get(attr, 0) + val

    hand_get = hand_counter.get
    attr_get = attr_counter.get
    card_attr_get = card_attr_index.get

    matched_count = 0

    for exact, wild in compiled_patterns:
        remaining_attrs = {}
        for card, count in exact.items():
            if hand_get(card, 0) < count:
                break
            card_attrs = card_attr_get(card)
            if card_attrs:
                for attr, val in card_attrs.items():
                    remaining_attrs[attr] = remaining_attrs.get(
                        attr, 0) - val * count
        else:
            for attr, count in wild.items():
                available = attr_get(attr, 0) + remaining_attrs.get(attr, 0)
                if available < count:
                    break
            else:
                matched_count += 1

    return matched_count


def run_test_hand_without_gambling(
    hand_checker: callable,
    hand: Sequence[int],
) -> HandTestResult:
    result = hand_checker(hand)
    return HandTestResult(
        matches_without_gambling=result,
        matches_with_gambling=result,
        rescued_with_gambling=0,
        useful_gambles=Counter(),
        gamble_seen=Counter(),
        gamble_attempted=0,
        gamble_failed=0,
        gamble_unplayable=0,
    )


def run_test_hand_with_gambling(
    hand_checker: callable,
    hand: Sequence[int],
    card_attr_index: CardAttrIndex,
    remaining_deck: list[int],
    gambling_cards: GamblingCards,
) -> HandTestResult:

    matches_without = hand_checker(hand)
    matches_with = matches_without
    rescued_with_gambling = 0

    useful_gambles: Counter[int] = Counter()
    gamble_seen: Counter[int] = Counter()
    gamble_attempted = 0
    gamble_failed = 0
    gamble_unplayable = 0

    if matches_without:
        return HandTestResult(
            matches_without_gambling=matches_without,
            matches_with_gambling=matches_with,
            rescued_with_gambling=rescued_with_gambling,
            useful_gambles=useful_gambles,
            gamble_seen=gamble_seen,
            gamble_attempted=gamble_attempted,
            gamble_failed=gamble_failed,
            gamble_unplayable=gamble_unplayable,
        )

    gamble_card = next((c for c in hand if c in gambling_cards), None)
    if gamble_card is None:
        return HandTestResult(
            matches_without_gambling=matches_without,
            matches_with_gambling=matches_with,
            rescued_with_gambling=rescued_with_gambling,
            useful_gambles=useful_gambles,
            gamble_seen=gamble_seen,
            gamble_attempted=gamble_attempted,
            gamble_failed=gamble_failed,
            gamble_unplayable=gamble_unplayable,
        )

    gamble_seen[gamble_card] += 1
    spec = gambling_cards[gamble_card]
    discard_requirements = spec.get("discard", [])

    if discard_requirements:
        discardable = [
            c for c in hand
            for field, value in discard_requirements
            if card_attr_index.get(c, {}).get((field, value), 0) > 0
        ]
        if not discardable:
            gamble_unplayable += 1
            return HandTestResult(
                matches_without_gambling=matches_without,
                matches_with_gambling=matches_with,
                rescued_with_gambling=rescued_with_gambling,
                useful_gambles=useful_gambles,
                gamble_seen=gamble_seen,
                gamble_attempted=gamble_attempted,
                gamble_failed=gamble_failed,
                gamble_unplayable=gamble_unplayable,
            )
    else:
        discardable = []

    new_hand = list(hand)
    new_hand.remove(gamble_card)
    num_to_draw = spec.get("draw", 0)
    if len(remaining_deck) < num_to_draw:
        gamble_unplayable += 1
        return HandTestResult(
            matches_without_gambling=matches_without,
            matches_with_gambling=matches_with,
            rescued_with_gambling=rescued_with_gambling,
            useful_gambles=useful_gambles,
            gamble_seen=gamble_seen,
            gamble_attempted=gamble_attempted,
            gamble_failed=gamble_failed,
            gamble_unplayable=gamble_unplayable,
        )

    gamble_attempted += 1
    drawn_cards = random.sample(remaining_deck, num_to_draw)
    new_hand.extend(drawn_cards)

    for c in new_hand:
        for field, value in discard_requirements:
            if card_attr_index.get(c, {}).get((field, value), 0) > 0:
                new_hand.remove(c)
                break
        else:
            continue
        break

    if hand_checker(new_hand):
        matches_with = True
        rescued_with_gambling = 1
        useful_gambles[gamble_card] += 1
    else:
        gamble_failed += 1

    return HandTestResult(
        matches_without_gambling=matches_without,
        matches_with_gambling=matches_with,
        rescued_with_gambling=rescued_with_gambling,
        useful_gambles=useful_gambles,
        gamble_seen=gamble_seen,
        gamble_attempted=gamble_attempted,
        gamble_failed=gamble_failed,
        gamble_unplayable=gamble_unplayable,
    )


def simple_consistency(
    deckcount: int,
    ratios: Sequence[int],
    names: Sequence[int],
    hand_tester: Callable,  # Returns HandTestResult
    num_hands: int = 1_000_000,
) -> ConsistencyResult:
    ratios = ratios.copy()
    names = names.copy()

    blanks = deckcount - sum(ratios)
    if blanks < 0:
        raise InvalidCardCountsError("Ratios add up to more than deck count")
    if blanks > 0:
        ratios.append(blanks)
        names.append(0)

    deck: list[int] = [int(card) for name, count in zip(
        names, ratios) for card in [name] * count]
    deckcount = len(deck)

    good_5 = good_6 = rescued_5 = rescued_6 = 0
    failed_gambles_5 = failed_gambles_6 = 0
    unplayable_gambles_5 = unplayable_gambles_6 = 0
    gamble_attempted_5 = gamble_attempted_6 = 0
    gamble_seen_5: Counter[int] = Counter()
    gamble_seen_6: Counter[int] = Counter()
    useful_gambles_5: Counter[int] = Counter()
    useful_gambles_6: Counter[int] = Counter()

    # New actionable metrics
    near_miss_counts: Counter[int] = Counter()
    blocking_card_counts: Counter[int] = Counter()
    ideal_hand_counts: Counter[int] = Counter()

    # Track per-hand matched patterns (raw counts)
    matched_pattern_counts_5: Counter[int] = Counter()
    matched_pattern_counts_6: Counter[int] = Counter()
    matched_pattern_counts_5_withgamble: Counter[int] = Counter()
    matched_pattern_counts_6_withgamble: Counter[int] = Counter()

    for _ in range(num_hands):
        hand5 = random.sample(deck, min(5, deckcount))
        remaining_deck_5 = deck.copy()
        for card in hand5:
            remaining_deck_5.remove(card)

        result5: HandTestResult = hand_tester(remaining_deck_5, hand5.copy())

        # Raw counts for pattern matches
        matched_pattern_counts_5[result5.matches_without_gambling] += 1
        total_matches_5 = result5.matches_without_gambling + result5.rescued_with_gambling
        matched_pattern_counts_5_withgamble[total_matches_5] += 1

        if result5.matches_without_gambling > 0:
            good_5 += 1
            for c in hand5:
                ideal_hand_counts[c] += 1
        else:
            for c in hand5:
                blocking_card_counts[c] += 1
            for c in set(deck) - set(hand5):
                near_miss_counts[c] += 1

        rescued_5 += result5.rescued_with_gambling
        failed_gambles_5 += result5.gamble_failed
        unplayable_gambles_5 += result5.gamble_unplayable
        gamble_attempted_5 += result5.gamble_attempted
        gamble_seen_5.update(result5.gamble_seen)
        useful_gambles_5.update(result5.useful_gambles)

        if deckcount >= 6:
            remaining_deck_for_6 = remaining_deck_5.copy()
            extra_card = random.choice(remaining_deck_for_6)
            hand6 = hand5 + [extra_card]
            remaining_deck_for_6.remove(extra_card)

            result6: HandTestResult = hand_tester(remaining_deck_for_6, hand6)

            matched_pattern_counts_6[result6.matches_without_gambling] += 1
            total_matches_6 = result6.matches_without_gambling + result6.rescued_with_gambling
            matched_pattern_counts_6_withgamble[total_matches_6] += 1

            if result6.matches_without_gambling > 0:
                good_6 += 1
                for c in hand6:
                    ideal_hand_counts[c] += 1
            else:
                for c in hand6:
                    blocking_card_counts[c] += 1
                for c in set(deck) - set(hand6):
                    near_miss_counts[c] += 1

            rescued_6 += result6.rescued_with_gambling
            failed_gambles_6 += result6.gamble_failed
            unplayable_gambles_6 += result6.gamble_unplayable
            gamble_attempted_6 += result6.gamble_attempted
            gamble_seen_6.update(result6.gamble_seen)
            useful_gambles_6.update(result6.useful_gambles)

    p5 = good_5 / num_hands
    p6 = good_6 / num_hands if deckcount >= 6 else 0.0
    p5_with_gambling = (good_5 + rescued_5) / num_hands
    p6_with_gambling = (good_6 + rescued_6) / \
        num_hands if deckcount >= 6 else 0.0

    return ConsistencyResult(
        num_hands=num_hands,
        p5=p5,
        p6=p6,
        p5_with_gambling=p5_with_gambling,
        p6_with_gambling=p6_with_gambling,
        rescued_5=rescued_5,
        rescued_6=rescued_6,
        useful_gambles_5=useful_gambles_5,
        useful_gambles_6=useful_gambles_6,
        gamble_seen_5=gamble_seen_5,
        gamble_seen_6=gamble_seen_6,
        gamble_attempted_5=gamble_attempted_5,
        gamble_attempted_6=gamble_attempted_6,
        failed_gambles_5=failed_gambles_5,
        failed_gambles_6=failed_gambles_6,
        unplayable_gambles_5=unplayable_gambles_5,
        unplayable_gambles_6=unplayable_gambles_6,
        near_miss_counts=near_miss_counts,
        blocking_card_counts=blocking_card_counts,
        ideal_hand_counts=ideal_hand_counts,
        matched_pattern_counts_5=matched_pattern_counts_5,
        matched_pattern_counts_6=matched_pattern_counts_6,
        matched_pattern_counts_5_withgamble=matched_pattern_counts_5_withgamble,
        matched_pattern_counts_6_withgamble=matched_pattern_counts_6_withgamble,
    )

File: frontend/src/services/calculator/calculator.ts
export type CardID = number
export type AttrKey = [string, string]
export type ExactPattern = Record<CardID, number>
export type WildPattern = Record<string, number> // tuple key as string
export type CompiledPattern = [ExactPattern, WildPattern]
export type CardAttrIndex = Record<CardID, Record<string, number>>

export interface Card {
    superType: string
    name: string
    race?: string
    attribute?: string
}

export type CardDatabase = Record<CardID, Card>
export type DiscardConstraint = [string, string]

export interface GambleCard {
    draw: number
    discard?: DiscardConstraint[]
}

export type GamblingCards = Record<CardID, GambleCard>

export interface HandTestResult {
    matches_without_gambling: number
    matches_with_gambling: number
    rescued_with_gambling: number
    useful_gambles: Map<CardID, number>
    gamble_seen: Map<CardID, number>
    gamble_attempted: number
    gamble_failed: number
    gamble_unplayable: number
}

export interface ConsistencyResult {
    num_hands: number
    p5: number
    p6: number
    p5_with_gambling: number
    p6_with_gambling: number
    rescued_5: number
    rescued_6: number
    useful_gambles_5: Map<CardID, number>
    useful_gambles_6: Map<CardID, number>
    gamble_seen_5: Map<CardID, number>
    gamble_seen_6: Map<CardID, number>
    gamble_attempted_5: number
    gamble_attempted_6: number
    failed_gambles_5: number
    failed_gambles_6: number
    unplayable_gambles_5: number
    unplayable_gambles_6: number
    near_miss_counts: Map<CardID, number>
    blocking_card_counts: Map<CardID, number>
    ideal_hand_counts: Map<CardID, number>
    matched_pattern_counts_5: Map<number, number>
    matched_pattern_counts_6: Map<number, number>
    matched_pattern_counts_5_withgamble: Map<number, number>
    matched_pattern_counts_6_withgamble: Map<number, number>
}

function counter<T>(iterable: T[]): Map<T, number> {
    const map = new Map<T, number>()
    for (const item of iterable) {
        map.set(item, (map.get(item) || 0) + 1)
    }
    return map
}

function randomIndex(max: number): number {
    return Math.floor(Math.random() * max)
}

export function hand_is_good(
    hand: CardID[],
    ideal_hands: (CardID[] | Map<CardID, number>)[],
): boolean {
    const hand_counter = counter(hand)

    const ideal_counters = ideal_hands.map((pattern) =>
        pattern instanceof Map ? pattern : counter(pattern as CardID[]),
    )

    return ideal_counters.some((pattern) =>
        Array.from(pattern.entries()).every(
            ([card, count]) => (hand_counter.get(card) || 0) >= count,
        ),
    )
}

export function hand_is_wild(
    hand: CardID[],
    compiled_patterns: CompiledPattern[],
    card_attr_index: CardAttrIndex,
): number {
    const hand_counter: Record<CardID, number> = {}
    const attr_counter: Record<string, number> = {}

    for (const c of hand) {
        hand_counter[c] = (hand_counter[c] || 0) + 1
    }

    for (const card of hand) {
        const attrs = card_attr_index[card]
        if (attrs) {
            for (const [attr, val] of Object.entries(attrs)) {
                attr_counter[attr] = (attr_counter[attr] || 0) + val
            }
        }
    }

    let matched_count = 0

    for (const [exact, wild] of compiled_patterns) {
        const remaining_attrs: Record<string, number> = {}

        let exact_match = true
        for (const [cardStr, count] of Object.entries(exact)) {
            const card = Number(cardStr)
            if ((hand_counter[card] || 0) < count) {
                exact_match = false
                break
            }
            const attrs = card_attr_index[card]
            if (attrs) {
                for (const [attr, val] of Object.entries(attrs)) {
                    remaining_attrs[attr] =
                        (remaining_attrs[attr] || 0) - val * count
                }
            }
        }
        if (!exact_match) continue

        let wild_match = true
        for (const [attr, count] of Object.entries(wild)) {
            if (
                (attr_counter[attr] || 0) + (remaining_attrs[attr] || 0) <
                count
            ) {
                wild_match = false
                break
            }
        }

        if (wild_match) matched_count += 1
    }

    return matched_count
}

export function run_test_hand_without_gambling(
    hand_checker: (hand: CardID[]) => number,
    hand: CardID[],
): HandTestResult {
    const result = hand_checker(hand)
    return {
        matches_without_gambling: result,
        matches_with_gambling: result,
        rescued_with_gambling: 0,
        useful_gambles: new Map(),
        gamble_seen: new Map(),
        gamble_attempted: 0,
        gamble_failed: 0,
        gamble_unplayable: 0,
    }
}

export function run_test_hand_with_gambling(
    hand_checker: (hand: CardID[]) => number,
    hand: CardID[],
    card_attr_index: CardAttrIndex,
    remaining_deck: CardID[],
    gambling_cards: GamblingCards,
): HandTestResult {
    let matches_without = hand_checker(hand)
    let matches_with = matches_without
    let rescued_with_gambling = 0

    const useful_gambles = new Map<CardID, number>()
    const gamble_seen = new Map<CardID, number>()
    let gamble_attempted = 0
    let gamble_failed = 0
    let gamble_unplayable = 0

    if (matches_without > 0) {
        return {
            matches_without_gambling: matches_without,
            matches_with_gambling: matches_with,
            rescued_with_gambling,
            useful_gambles,
            gamble_seen,
            gamble_attempted,
            gamble_failed,
            gamble_unplayable,
        }
    }

    const gamble_card = hand.find((c) => gambling_cards[c] !== undefined)
    if (gamble_card === undefined) {
        return {
            matches_without_gambling: matches_without,
            matches_with_gambling: matches_with,
            rescued_with_gambling,
            useful_gambles,
            gamble_seen,
            gamble_attempted,
            gamble_failed,
            gamble_unplayable,
        }
    }

    gamble_seen.set(gamble_card, (gamble_seen.get(gamble_card) || 0) + 1)

    const spec = gambling_cards[gamble_card]
    const discard_requirements = spec.discard ?? []

    let discardable: CardID[] = []

    if (discard_requirements.length > 0) {
        discardable = hand.filter((c) =>
            discard_requirements.some(
                ([field, value]) =>
                    card_attr_index[c]?.[`${field},${value}`] > 0,
            ),
        )
        if (discardable.length === 0) {
            gamble_unplayable += 1
            return {
                matches_without_gambling: matches_without,
                matches_with_gambling: matches_with,
                rescued_with_gambling,
                useful_gambles,
                gamble_seen,
                gamble_attempted,
                gamble_failed,
                gamble_unplayable,
            }
        }
    }

    const new_hand = hand.slice()
    new_hand.splice(new_hand.indexOf(gamble_card), 1)

    const num_to_draw = spec.draw ?? 0
    if (remaining_deck.length < num_to_draw) {
        gamble_unplayable += 1
        return {
            matches_without_gambling: matches_without,
            matches_with_gambling: matches_with,
            rescued_with_gambling,
            useful_gambles,
            gamble_seen,
            gamble_attempted,
            gamble_failed,
            gamble_unplayable,
        }
    }

    gamble_attempted += 1

    // Draw random cards without replacement
    const drawn_cards: CardID[] = []
    const temp_deck = remaining_deck.slice()
    for (let i = 0; i < num_to_draw; i++) {
        const idx = randomIndex(temp_deck.length)
        drawn_cards.push(temp_deck[idx])
        temp_deck.splice(idx, 1)
    }

    new_hand.push(...drawn_cards)

    // Remove discardable cards
    for (const c of new_hand) {
        if (
            discard_requirements.some(
                ([field, value]) =>
                    card_attr_index[c]?.[`${field},${value}`] > 0,
            )
        ) {
            new_hand.splice(new_hand.indexOf(c), 1)
            break
        }
    }

    if (hand_checker(new_hand) > 0) {
        matches_with = 1
        rescued_with_gambling = 1
        useful_gambles.set(
            gamble_card,
            (useful_gambles.get(gamble_card) || 0) + 1,
        )
    } else {
        gamble_failed += 1
    }

    return {
        matches_without_gambling: matches_without,
        matches_with_gambling: matches_with,
        rescued_with_gambling,
        useful_gambles,
        gamble_seen,
        gamble_attempted,
        gamble_failed,
        gamble_unplayable,
    }
}

export function simple_consistency(
    deckcount: number,
    ratios: number[],
    names: number[],
    hand_tester: (deck: CardID[], hand: CardID[]) => HandTestResult,
    num_hands = 1_000_000,
): ConsistencyResult {
    ratios = [...ratios]
    names = [...names]

    const blanks = deckcount - ratios.reduce((a, b) => a + b, 0)
    if (blanks < 0) throw new Error('Ratios add up to more than deck count')
    if (blanks > 0) {
        ratios.push(blanks)
        names.push(0)
    }

    const deck: CardID[] = []
    for (let i = 0; i < names.length; i++) {
        for (let j = 0; j < ratios[i]; j++) {
            deck.push(names[i])
        }
    }

    deckcount = deck.length

    let good_5 = 0,
        good_6 = 0,
        rescued_5 = 0,
        rescued_6 = 0,
        failed_gambles_5 = 0,
        failed_gambles_6 = 0,
        unplayable_gambles_5 = 0,
        unplayable_gambles_6 = 0,
        gamble_attempted_5 = 0,
        gamble_attempted_6 = 0

    const gamble_seen_5 = new Map<CardID, number>()
    const gamble_seen_6 = new Map<CardID, number>()
    const useful_gambles_5 = new Map<CardID, number>()
    const useful_gambles_6 = new Map<CardID, number>()

    const near_miss_counts = new Map<CardID, number>()
    const blocking_card_counts = new Map<CardID, number>()
    const ideal_hand_counts = new Map<CardID, number>()
    const matched_pattern_counts_5 = new Map<number, number>()
    const matched_pattern_counts_6 = new Map<number, number>()
    const matched_pattern_counts_5_withgamble = new Map<number, number>()
    const matched_pattern_counts_6_withgamble = new Map<number, number>()

    for (let n = 0; n < num_hands; n++) {
        const hand5: CardID[] = []
        const temp_deck = deck.slice()
        for (let i = 0; i < Math.min(5, deckcount); i++) {
            const idx = randomIndex(temp_deck.length)
            hand5.push(temp_deck[idx])
            temp_deck.splice(idx, 1)
        }

        const result5 = hand_tester(temp_deck, hand5.slice())

        matched_pattern_counts_5.set(
            result5.matches_without_gambling,
            (matched_pattern_counts_5.get(result5.matches_without_gambling) ||
                0) + 1,
        )

        const total_matches_5 =
            result5.matches_without_gambling + result5.rescued_with_gambling
        matched_pattern_counts_5_withgamble.set(
            total_matches_5,
            (matched_pattern_counts_5_withgamble.get(total_matches_5) || 0) + 1,
        )

        if (result5.matches_without_gambling > 0) {
            good_5 += 1
            for (const c of hand5)
                ideal_hand_counts.set(c, (ideal_hand_counts.get(c) || 0) + 1)
        } else {
            for (const c of hand5)
                blocking_card_counts.set(
                    c,
                    (blocking_card_counts.get(c) || 0) + 1,
                )
            for (const c of deck.filter((d) => !hand5.includes(d))) {
                near_miss_counts.set(c, (near_miss_counts.get(c) || 0) + 1)
            }
        }

        rescued_5 += result5.rescued_with_gambling
        failed_gambles_5 += result5.gamble_failed
        unplayable_gambles_5 += result5.gamble_unplayable
        gamble_attempted_5 += result5.gamble_attempted

        result5.gamble_seen.forEach((v, k) =>
            gamble_seen_5.set(k, (gamble_seen_5.get(k) || 0) + v),
        )
        result5.useful_gambles.forEach((v, k) =>
            useful_gambles_5.set(k, (useful_gambles_5.get(k) || 0) + v),
        )

        if (deckcount >= 6) {
            const temp_deck_for_6 = temp_deck.slice()
            const extra_card = temp_deck_for_6[randomIndex(temp_deck.length)]
            const hand6 = hand5.concat(extra_card)
            temp_deck_for_6.splice(temp_deck_for_6.indexOf(extra_card), 1)

            const result6 = hand_tester(temp_deck_for_6, hand6)

            matched_pattern_counts_6.set(
                result6.matches_without_gambling,
                (matched_pattern_counts_6.get(
                    result6.matches_without_gambling,
                ) || 0) + 1,
            )

            const total_matches_6 =
                result6.matches_without_gambling + result6.rescued_with_gambling
            matched_pattern_counts_6_withgamble.set(
                total_matches_6,
                (matched_pattern_counts_6_withgamble.get(total_matches_6) ||
                    0) + 1,
            )

            if (result6.matches_without_gambling > 0) {
                good_6 += 1
                for (const c of hand6)
                    ideal_hand_counts.set(
                        c,
                        (ideal_hand_counts.get(c) || 0) + 1,
                    )
            } else {
                for (const c of hand6)
                    blocking_card_counts.set(
                        c,
                        (blocking_card_counts.get(c) || 0) + 1,
                    )
                for (const c of deck.filter((d) => !hand6.includes(d))) {
                    near_miss_counts.set(c, (near_miss_counts.get(c) || 0) + 1)
                }
            }

            rescued_6 += result6.rescued_with_gambling
            failed_gambles_6 += result6.gamble_failed
            unplayable_gambles_6 += result6.gamble_unplayable
            gamble_attempted_6 += result6.gamble_attempted

            result6.gamble_seen.forEach((v, k) =>
                gamble_seen_6.set(k, (gamble_seen_6.get(k) || 0) + v),
            )
            result6.useful_gambles.forEach((v, k) =>
                useful_gambles_6.set(k, (useful_gambles_6.get(k) || 0) + v),
            )
        }
    }

    const p5 = good_5 / num_hands
    const p6 = deckcount >= 6 ? good_6 / num_hands : 0.0
    const p5_with_gambling = (good_5 + rescued_5) / num_hands
    const p6_with_gambling =
        deckcount >= 6 ? (good_6 + rescued_6) / num_hands : 0.0

    return {
        num_hands,
        p5,
        p6,
        p5_with_gambling,
        p6_with_gambling,
        rescued_5,
        rescued_6,
        useful_gambles_5,
        useful_gambles_6,
        gamble_seen_5,
        gamble_seen_6,
        gamble_attempted_5,
        gamble_attempted_6,
        failed_gambles_5,
        failed_gambles_6,
        unplayable_gambles_5,
        unplayable_gambles_6,
        near_miss_counts,
        blocking_card_counts,
        ideal_hand_counts,
        matched_pattern_counts_5,
        matched_pattern_counts_6,
        matched_pattern_counts_5_withgamble,
        matched_pattern_counts_6_withgamble,
    }
}
